protocols/autonat/src/v2/client/handler/dial_back.rs

```diff
@@ -0,0 +1,142 @@
+        SubstreamProtocol::new(ReadyUpgrade::new(DIAL_BACK_PROTOCOL), ())
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<
+        ConnectionHandlerEvent<Self::OutboundProtocol, Self::OutboundOpenInfo, Self::ToBehaviour>,
+    > {
+        match self.inbound.poll_next_unpin(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(None) => Poll::Pending,
+            Poll::Ready(Some(Err(err))) => {
+                tracing::debug!("Stream timed out: {err}");
+                Poll::Pending
+            }
+            Poll::Ready(Some(Ok(Err(err)))) => {
+                tracing::debug!("Dial back handler failed with: {err:?}");
+                Poll::Pending
+            }
+            Poll::Ready(Some(Ok(Ok(incoming_nonce)))) => {
+                Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(incoming_nonce))
+            }
+        }
+    }
+
+    fn on_behaviour_event(&mut self, _event: Self::FromBehaviour) {}
+
+    fn on_connection_event(
+        &mut self,
+        event: ConnectionEvent<
+            Self::InboundProtocol,
+            Self::OutboundProtocol,
+            Self::InboundOpenInfo,
+            Self::OutboundOpenInfo,
+        >,
+    ) {
+        match event {
+            ConnectionEvent::FullyNegotiatedInbound(FullyNegotiatedInbound {
+                protocol, ..
+            }) => {
+                if self.inbound.try_push(perform_dial_back(protocol)).is_err() {
+                    tracing::warn!("Dial back request dropped, too many requests in flight");
+                }
+            }
+            ConnectionEvent::ListenUpgradeError(ListenUpgradeError { error, .. }) => {
+                void::unreachable(error);
+            }
+            _ => {}
+        }
+    }
+}
+
+struct State {
+    stream: libp2p_swarm::Stream,
+    oneshot: Option<oneshot::Receiver<io::Result<()>>>,
+}
+
+#[derive(Debug)]
+pub struct IncomingNonce {
+    pub nonce: Nonce,
+    pub sender: oneshot::Sender<io::Result<()>>,
+}
+
+fn perform_dial_back(
+    stream: libp2p_swarm::Stream,
+) -> impl futures::Stream<Item = io::Result<IncomingNonce>> {
+    let state = State {
+        stream,
+        oneshot: None,
+    };
+    futures::stream::unfold(state, |mut state| async move {
+        if let Some(ref mut receiver) = state.oneshot {
+            match receiver.await {
+                Ok(Ok(())) => {}
+                Ok(Err(e)) => return Some((Err(e), state)),
+                Err(_) => {
+                    return Some((
+                        Err(io::Error::new(io::ErrorKind::Other, "Sender got cancelled")),
+                        state,
+                    ));
+                }
+            }
+            if let Err(e) = protocol::dial_back_response(&mut state.stream).await {
+                return Some((Err(e), state));
+            }
+            if let Err(e) = state.stream.close().await {
+                return Some((Err(e), state));
+            }
```

Comment:
```suggestion
```

You are already closing the stream in `dial_back_response`.

Reply:
thanks, I overlooked that.